src/migrations/rebase.rs

```diff
@@ -0,0 +1,286 @@
+}
+
+struct RebaseMigration<'a> {
+    key: MigrationKey,
+    migration: &'a DBMigration,
+    parent: Option<&'a str>,
+    kind: RebaseMigrationKind
+}
+
+impl MigrationToText for RebaseMigration<'_> {
+    type StatementsIter<'a> = std::iter::Once<&'a String> where Self: 'a;
+
+    fn key(&self) -> &MigrationKey {
+        &self.key
+    }
+
+    fn parent(&self) -> anyhow::Result<&str> {
+        if let Some(parent) = self.parent {
+            return Ok(parent)
+        }
+
+        if self.migration.parent_names.is_empty() {
+            return Ok("initial");
+        }
+
+        Ok(&self.migration.parent_names[0])
+    }
+
+    fn id(&self) -> anyhow::Result<&str> {
+        Ok(&self.migration.name)
+    }
+
+    fn statements<'a>(&'a self) -> Self::StatementsIter<'a> {
+        std::iter::once(&self.migration.script)
+    }
+}
+
+pub struct RebaseMigrations {
+    /// initial..base : the commonly shared migrations between both 'source' and 'target'
+    base_migrations: IndexMap<String, DBMigration>,
+    /// base..source : the migrations that are unique to 'source'
+    source_migrations: IndexMap<String, DBMigration>,
+    /// base..target : the migrations that are unique to 'target'
+    target_migrations: IndexMap<String, DBMigration>,
+}
+
+impl RebaseMigrations {
+    pub fn print_status(&self) {
+        let last_common = self.base_migrations.last().map(|v| v.0.as_str()).unwrap_or("initial").green();
+
+        let format_migration_on_length = |c: usize| {
+             if c > 1 {"migrations"} else {"migration"}
+        };
+
+        eprintln!("Last common migration is {}", last_common);
+        eprintln!(
+            "Since then, there are:\n- {} new {} on the target branch,\n- {} {} to rebase",
+            self.target_migrations.len().to_string().green(), format_migration_on_length(self.target_migrations.len()),
+            self.source_migrations.len().to_string().green(), format_migration_on_length(self.source_migrations.len())
+        );
+    }
+
+    fn flatten(&self) -> anyhow::Result<Vec<RebaseMigration>> {
+        let mut result = Vec::new();
+        let mut key_index = 0;
+
+        let mut next_key = || -> MigrationKey {
+            key_index += 1;
+            MigrationKey::Index(key_index)
+        };
+
+        for (_, migration) in &self.base_migrations {
+            result.push(RebaseMigration {
+                key: next_key(),
+                parent: None,
+                migration,
+                kind: RebaseMigrationKind::Base,
+            });
+        }
+
+        if self.source_migrations.len() > 0 {
+            let (_, first_migration) = self.source_migrations.first().unwrap();
+            let base_last = result.last();
+
+            result.push(RebaseMigration {
+                key: next_key(),
+                migration: first_migration,
+                parent: base_last.map(|v| v.migration.name.as_str()),
+                kind: RebaseMigrationKind::Source,
+            });
```

Comment:
Is there a reason for special-casing the first migration?

Reply:
Yes, since the 3 sets need to be "joined" by the parent id for the rebase to work properly, thats what `parent` override does

Reply:
Ahh ok, I see now. It is `parent_override`! Can you rename it to `parent_override` (only if you decide against my other comment about setting all `parent`s)